Adjust bootstrap sequence

droberts195 opened the pull request.

tlrx commented:
I don't have the full context but I think it could be nice to also test this using packaging tests as a follow up PR. Similar to the `jvm-example` plugin we could have a `spawn-plugin` that does similar things like writing hello/goodbye files but packaging test would help to test security as well as that the spawn process is related to the elasticsearch process (PPID) as well as shut down when elasticsearch is stopped.

s1monw commented on qa/evil-tests/src/test/java/org/elasticsearch/bootstrap/EvilSpawnerTests.java:
```diff
@@ -0,0 +1,164 @@
+
+        try {
+            // Give the program time to start up
+            Thread.sleep(500);
```
instead of adding a sleep here you can do this

```Java
assertBusy(() -> {
         List<OutputStream> stdinReferences = spawner.getStdinReferences();
           // 1 because there should only be a reference in the list for the plugin that had the controller daemon, not the other plugin
          assertEquals(1, stdinReferences.size());
          assertTrue(Files.isRegularFile(helloFile));
          assertFalse(Files.exists(goodbyeFile));
          List<String> lines = Files.readAllLines(helloFile, StandardCharsets.UTF_8);
          assertFalse(lines.isEmpty());
          assertEquals("Hello, world!", lines.get(0));
});
spawner.close();
assertBusy(() -> {
         // Give the program time to exit
         assertTrue(Files.isRegularFile(goodbyeFile));
         lines = Files.readAllLines(goodbyeFile, StandardCharsets.UTF_8);
         assertFalse(lines.isEmpty());
         assertEquals("Goodbye, world!", lines.get(0));
});
```

`assertBusy` will try to execute the code and pause for an quadratically growing amount of time between retries. if it fails after retrying N times it will just rethrow the exception it's much more robust and will also pass on slowish machines

s1monw commented on qa/evil-tests/src/test/java/org/elasticsearch/bootstrap/EvilSpawnerTests.java:
```diff
@@ -0,0 +1,164 @@
+    private static final String HELLO_FILE_NAME = JvmInfo.jvmInfo().pid() + "hello.txt";
+    private static final String GOODBYE_FILE_NAME = JvmInfo.jvmInfo().pid() + "goodbye.txt";
+
+    private static final String CONTROLLER_SOURCE = "\n"
```
can this just be something like this

```BASH
#!/bin/bash
echo "foo bar" >> ${TMPDIR}/somefile.txt
```
instead?

s1monw commented on core/src/main/java/org/elasticsearch/bootstrap/Spawner.java:
```diff
@@ -0,0 +1,129 @@
+     * will remain connected to this JVM via its stdin, stdout and stderr, but the
+     * references to these streams are not available to code outside this class.
+     */
+    private void spawnNativePluginController(Environment environment, Path spawnPath) throws IOException {
```
can we make the environment variables passed on to this configurable ie. as ctor arguments? I also wonder if the variable should be prefixed with `ES_`

s1monw replied:
or rather pass it to `spawnNativePluginControllers` and change it's signature to `spawnNativePluginControllers(Path pluginPath, Map<String,String> env)` I don't think we need to depend on `Environment`

s1monw commented on core/src/main/java/org/elasticsearch/bootstrap/Spawner.java:
```diff
@@ -0,0 +1,129 @@
+    public void close() throws IOException {
+        for (OutputStream stream : stdinReferences) {
+            stream.close();
+        }
```
lets call `stdinReferences.clear()` after we closed all of them. I also think you should use `IOUtils.close(stdinReferences)` instead, it will close all references even if one close call throws an exception.

droberts195 commented:
I changed the test that spawns a controller program to use a shell script rather than a native program, and that simplified things a lot.

I also changed the shutdown to explicitly kill the daemon processes rather than trusting them to exit when their stdin gets an end-of-file.

Finally, to get around the issue of other evil tests installing seccomp, I moved the test that actually spawns a controller into a completely separate gradle project so it runs in its own JVM.  (If we ever have any other tests that need to run without the ESTestCase bootstrapping, they could also go in that same project.  Alternatively, if there was a better place to put it please let me know.)

s1monw commented on qa/no-bootstrap-tests/src/test/java/org/elasticsearch/bootstrap/SpawnerNoBootstrapTests.java:
```diff
@@ -0,0 +1,110 @@
+            + "read SOMETHING\n";
+
+    public void testControllerSpawn() throws IOException, InterruptedException {
+        if (Constants.WINDOWS) {
```
please use `assumeFalse(Constants.WINDOWS);`

s1monw commented on core/src/main/java/org/elasticsearch/bootstrap/Spawner.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    public void close() throws IOException {
+        for (Process process : processes) {
```
maybe like this:
```Java

try {
  IOUtils.close(() -> processes.stream().map(s -> (Closeable)s::destroy).iterator());
} finally {
  processes.clear();
}

s1monw commented on core/src/main/java/org/elasticsearch/bootstrap/Spawner.java:
```diff
@@ -0,0 +1,125 @@
+
+        // The only environment variable passes on the path to the temporary directory
+        pb.environment().clear();
+        pb.environment().put(TMP_ENVVAR, tmpPath.toString());
```
the reason why I suggested to make it configurable is that we could pass in our own values in tests that's all... not a big deal

s1monw commented on core/src/main/java/org/elasticsearch/bootstrap/Spawner.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    List<Process> getProcesses() {
+        return processes;
```
nit: `return Collections.unmodifiableList(processes);`

droberts195 commented:
Thanks for all the help @s1monw. I have made the 3 changes you asked for. I'll wait for @rjernst to have a look before merging.

s1monw commented:
@droberts195 thank you for all the iterations!
